### PR TITLE
mohammad darwesh homework js3 week3

### PR DESCRIPTION
**Pull request by mohammadnaefaldarwesh**

### CODE REVIEW

File: homework/index.js [hunk 1]
```diff
@@ -29,16 +14,110 @@
...
+    const RepositoryName = createAndAppend('li', repositoryInformation, { text: '' });
+    const description = createAndAppend('li', repositoryInformation, { text: '' });
+    const forks = createAndAppend('li', repositoryInformation, { text: '' });
+    const updat = createAndAppend('li', repositoryInformation, { text: '' });
+
+    for (let i = 0; i < jsonFile.length; i++) {
+      createAndAppend('option', repositories, { text: jsonFile[i].name, value: i });
+    }
+
+    async function displayRepoInformation() {
```
Comment (yash-kapila) [inline homework/index.js]: Another thing here - even though the logic works fine here, I think there is still some room for improvement. For example, the `makeApp` function is still quite big and is doing more than one thing. It is responsible for creating all elements for repositories, displays information in those created elements and then fetches contributors information.

I think we can split this function a bit more. For example, the parent function can be responsible for creating all elements and the dropdown for repositories. The `displayRepoInformation` function can be moved out of the `makeApp` function. As far as the `repositories.value`, `repositoryName`, `description` and other parameters are concerned, they can be passed on to this function as parameters.

Drop me a message on Slack if you have any questions around it.  🙂

File: homework/index.js [hunk 2]
```diff
@@ -29,16 +14,110 @@
...
+      class: 'alert-error',
+    });
+  }
+
+  const renderContributors = contributor => {
+    const contributions = document.getElementById('contributions');
+    contributions.innerHTML = 'Contributors';
+    if (contributor && contributor.length > 0) {
+      for (let contribution = 0; contribution < contributor.length; contribution++) {
+        const logi = contributor[contribution].login;
```
Comment (yash-kapila) [inline homework/index.js]: A spelling mistake here as well 😉

File: homework/Contributor.js [hunk 3]
```diff
@@ -13,7 +13,21 @@ class Contributor {
...
+    const img = this.contributor.avatar_url;
+    const link = this.contributor.html_url;
+    const contribute = this.contributor.contributions;
+    const li = Util.createAndAppend('li', container, { class: 'right-li' });
+
+    const contributorsList = `<a target=_blank href=${link}><img src=${img}><br>${logi}<br>Contributions: ${contribute}</a>`;
+
+    li.innerHTML += contributorsList;
+
+    li.addEventListener('keyup', contributor => {
```
Comment (yash-kapila) [inline homework/Contributor.js]: Hi Darwesh - may I ask why did you add this event listener here?

File: homework/Repository.js [hunk 4]
```diff
@@ -13,15 +13,36 @@ class Repository {
...
+    }
   }
 
   /**
    * Returns an array of contributors as a promise
    */
-  fetchContributors() {
-    return Util.fetchJSON(this.repository.contributors_url);
+  async fetchContributors() {
+    const reso = await Util.fetchJSON(this.repository.contributors_url);
```
Comment (yash-kapila) [inline homework/Repository.js]: Umm.. I don't think this is the best name for a variable here. 🙂

File: homework/App.js [hunk 5]
```diff
@@ -55,7 +71,10 @@ class App {
       const leftDiv = Util.createAndAppend('div', container);
       const rightDiv = Util.createAndAppend('div', container);
 
-      const contributorList = Util.createAndAppend('ul', rightDiv);
+      const contributorList = Util.createAndAppend('ul', rightDiv, {
+        text: 'Contributors',
+        class: 'rightDiv',
```
Comment (yash-kapila) [inline homework/App.js]: `kebab-case` instead of `camelCase` for class attributes please 🙂

File: homework/Repository.js [hunk 6]
```diff
@@ -13,15 +13,36 @@ class Repository {
...
+    const forks = Util.createAndAppend('li', repositoryInformation, { text: '' });
+    const update = Util.createAndAppend('li', repositoryInformation, { text: '' });
+
+    RepositoryName.innerHTML = `Repository:<a target=_blank href='${this.repository.html_url}'> ${
+      this.repository.name
+    }</a>`;
+
+    description.innerHTML = `Description: ${this.repository.description}`;
+    forks.innerHTML = `Forks: ${this.repository.forks}`;
+    update.innerHTML = `Updated: ${this.repository.updated_at}`;
```
Comment (yash-kapila) [inline homework/Repository.js]: Another thing here please - the date which is retrieved from the GitHub API comes in a `ISO 8601` format i.e `2018-04-06T17:18:06Z`. This can be shown in a more readable format like `dd/mm/yyyy`. For this, I would suggest that you explore the `Date` object of JavaScript and see if you can make this appear on UI in `dd/mm/yyyy` format - https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Date

File: homework/index.js [hunk 7]
```diff
@@ -29,16 +14,110 @@
...
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-      } else {
-        createAndAppend('pre', root, { text: JSON.stringify(data, null, 2) });
+  function showError(err) {
+    const root = document.getElementById('root');
+
+    createAndAppend('div', root, {
+      text: 'HYF Repositories',
+      class: 'HYF-Repositories',
```
[the remark this answers is not in the document]
Comment (mohammadnaefaldarwesh) [reply homework/index.js]: Done !

File: homework/index.js [hunk 8]
```diff
@@ -29,16 +14,110 @@
...
+        const contribute = contributor[contribution].contributions;
+        const contributorsList = `<li class='right-li'><a target=_blank href=${link}><img src=${img}><br>${logi}<br>Contributions: ${contribute}</a></li>`;
+
+        contributions.innerHTML += contributorsList;
       }
+    } else {
+      contributions.innerHTML += '<li>No information available</li>';
+    }
+  };
+  function makeApp(jsonFile) {
```
[the remark this answers is not in the document]
Comment (mohammadnaefaldarwesh) [reply homework/index.js]: Done !

File: homework/index.js [hunk 9]
```diff
@@ -29,16 +14,110 @@
...
+    jsonFile.sort((a, b) => a.name.localeCompare(b.name));
+    const root = document.getElementById('root');
+    const hyfRepositories = createAndAppend('div', root, {
+      class: 'HYF-Repositories',
+      id: 'HYF-Repositories',
```
[the remark this answers is not in the document]
Comment (mohammadnaefaldarwesh) [reply homework/index.js]: Done !

File: homework/index.js [hunk 10]
```diff
@@ -29,16 +14,110 @@
...
     });
+    const repositoryInformation = createAndAppend('ul', informationBox, {
+      id: 'repository-information',
+    });
+    const contributions = createAndAppend('ul', informationBox, {
+      text: 'Contributors',
+      id: 'contributions',
+      class: 'right-ul',
+    });
+    const RepositoryName = createAndAppend('li', repositoryInformation, { text: '' });
```
[the remark this answers is not in the document]
Comment (mohammadnaefaldarwesh) [reply homework/index.js]: Done !

File: homework/index.js (hunk 2 above)
Comment (mohammadnaefaldarwesh) [reply homework/index.js]: Done !

File: homework/Repository.js (hunk 4 above)
Comment (mohammadnaefaldarwesh) [reply homework/Repository.js]: Done !

File: homework/App.js (hunk 5 above)
Comment (mohammadnaefaldarwesh) [reply homework/App.js]: Done !

File: homework/Repository.js (hunk 6 above)
Comment (mohammadnaefaldarwesh) [reply homework/Repository.js]: Done !

File: homework/Contributor.js (hunk 3 above)
Comment (mohammadnaefaldarwesh) [reply homework/Contributor.js]: Hello @yash-kapila ,
Thank you for your feedback and your useful advises.
I add this event listener here to do this part from the homework:

 (a user should be able to navigate through all interactive elements using the keyboard (e.g., using the Tab key). Pressing Enter on such an element should be equivalent to clicking the mouse.)

I just check how to do it on google and i copy and past the code and it is work just fine .
Is there any wrong with it ?
Best regards,

Comment (yash-kapila) [reply homework/Contributor.js]: Ah no. There isn't anything wrong with it but keyboard navigation would work in this case even without it. Can you check? 🙂 

If you have anchor tags on the page, then browser automatically moves to them as you navigate through the page using keyboard.

**Comment by wouterkleijn** [comment]
Thanks a lot @mohammadnaefaldarwesh for addressing the feedback given by @yash-kapila Great job!
